File: apps/api/schemas/approvals.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class ApprovalResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: UUID
    workspace_id: UUID
    run_id: UUID
    agent_version_id: UUID
    logical_action_id: str
    tool_revision_id: UUID | None
    tool_identity: str
    canonical_arguments: dict[str, Any]
    canonical_args_hash: str
    decision_status: str
    execution_status: str
    requested_by: UUID
    decided_by: UUID | None
    decided_at: datetime | None
    claimed_at: datetime | None
    executed_at: datetime | None
    expires_at: datetime | None
    failure_code: str | None
    safe_failure_message: str | None
    safe_result: dict[str, Any] | None
    execution_attempt_count: int
    created_at: datetime
    updated_at: datetime


class ApprovalDecisionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    approval: ApprovalResponse
    run_id: UUID
    run_status: str

File: apps/web/lib/approvals.ts
import { apiRequest, ApiError } from "./api-client";

export { ApiError as ApprovalApiError };

export type Approval = {
  id: string;
  workspace_id: string;
  run_id: string;
  agent_version_id: string;
  logical_action_id: string;
  tool_revision_id: string | null;
  tool_identity: string;
  canonical_arguments: Record<string, unknown>;
  canonical_args_hash: string;
  decision_status: string;
  execution_status: string;
  requested_by: string;
  decided_by: string | null;
  decided_at: string | null;
  claimed_at: string | null;
  executed_at: string | null;
  expires_at: string | null;
  failure_code: string | null;
  safe_failure_message: string | null;
  safe_result: Record<string, unknown> | null;
  execution_attempt_count: number;
  created_at: string;
  updated_at: string;
};

export type ApprovalDecision = {
  approval: Approval;
  run_id: string;
  run_status: string;
};

export function listApprovals(workspaceId: string, accessToken: string): Promise<Approval[]> {
  return apiRequest<Approval[]>(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/approvals`,
    accessToken,
  );
}

export function decideApproval(
  workspaceId: string,
  approvalId: string,
  decision: "approve" | "deny",
  accessToken: string,
): Promise<ApprovalDecision> {
  return apiRequest<ApprovalDecision>(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/approvals/${encodeURIComponent(approvalId)}/${decision}`,
    accessToken,
    { method: "POST", body: {} },
  );
}
